### PR TITLE
Bug 1262981 - Filter subtests in graph view and show link when subtests are present

### PR DESCRIPTION
**Pull request by crosscent**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1560)

### CODE REVIEW

File: ui/js/models/perf/alerts.js [hunk 1]
```diff
@@ -34,6 +34,33 @@ treeherder.factory('PhAlerts', [
...
+                for (var key in associativeArray) {
+                    if (associativeArray.hasOwnProperty(key)) {
+                        parameters.push(key + '=' + encodeURIComponent(associativeArray[key]));
+                    }
+                }
+                return parameters.join('&');
+            }
+
+            var endpoint = '#/comparesubtest';
+            var urlParameters = {};
```
Comment (wlach) [inline ui/js/models/perf/alerts.js]: There's a slightly more concise way of expressing this:

```js
var urlParameters = {
  framework: alertSummary.framework,
  originalRepository: alertSummary.repository,
  originalSignature: this.series_signature.signature_hash,
  ...
}
```

File: ui/js/models/perf/alerts.js [hunk 2]
```diff
@@ -34,6 +34,33 @@ treeherder.factory('PhAlerts', [
 
             return url;
         };
+        Alert.prototype.getSubtestsURL = function(alertSummary) {
+            function formURLParameters(associativeArray) {
```
Comment (wlach) [inline ui/js/models/perf/alerts.js]: associativeArray -> params

File: ui/js/models/perf/alerts.js [hunk 3]
```diff
@@ -34,6 +34,33 @@ treeherder.factory('PhAlerts', [
...
+                var parameters = [];
+                for (var key in associativeArray) {
+                    if (associativeArray.hasOwnProperty(key)) {
+                        parameters.push(key + '=' + encodeURIComponent(associativeArray[key]));
+                    }
+                }
+                return parameters.join('&');
+            }
```
Comment (wlach) [inline ui/js/models/perf/alerts.js]: It's probably not necessary to define a function for this.

Something like this should work, if params is just an object:

```js
_.map(params, function(v, k) { return k + "=" + v; }).join('&');
```

File: ui/js/models/perf/alerts.js [hunk 4]
```diff
@@ -34,6 +34,24 @@ treeherder.factory('PhAlerts', [
...
+            var endpoint = '#/comparesubtest';
+            var urlParameters = {
+                framework: alertSummary.framework,
+                originalProject: alertSummary.repository,
+                originalSignature: this.series_signature.signature_hash,
+                newProject: alertSummary.repository,
+                newSignature: this.series_signature.signature_hash,
+            };
+            if (alertSummary.prevResultSetMetadata) {
+                urlParameters['originalRevision'] = alertSummary.prevResultSetMetadata.revision;
```
Comment (wlach) [inline ui/js/models/perf/alerts.js]: Could you update this to `urlParameters.originalRevision`? Ditto with the statement just below.

File: ui/js/controllers/perf/graphs.js [hunk 5]
```diff
@@ -1054,7 +1054,8 @@ perf.controller('TestChooserCtrl', function($scope, $uibModalInstance, $http,
                       subtests: $scope.includeSubtests ? 1 : 0 }).then(function(seriesList) {
                           $scope.unselectedTestList = _.sortBy(
                               _.filter(seriesList,
-                                       { platform: $scope.selectedPlatform }), 'name');
+                                       { platform: $scope.selectedPlatform,
+                                         hasSubtests: true }), 'name');
```
Comment (wlach) [inline ui/js/controllers/perf/graphs.js]: Sorry I missed this earlier -- this change doesn't look correct?

File: ui/partials/perf/alertsctrl.html [hunk 6]
```diff
@@ -125,7 +125,7 @@
           <span class="result-links">
             <a href="{{alert.getGraphsURL(alertSummary.resultSetMetadata.timeRange, alertSummary.repository, alertSummary.framework)}}" target="_blank">graph</a>
             <span> · </span>
```
Comment (wlach) [inline ui/partials/perf/alertsctrl.html]: This should be:

```html
 <span ng-if="alert.series_signature.has_subtests"> · </span>
```

File: ui/partials/perf/alertsctrl.html [hunk 7]
```diff
@@ -125,7 +125,7 @@
...
-            <a href="#/comparesubtest?originalProject={{alertSummary.repository}}&originalRevision={{alertSummary.prevResultSetMetadata.revision}}&newProject={{alertSummary.repository}}&newRevision={{alertSummary.resultSetMetadata.revision}}&originalSignature={{alert.series_signature.signature_hash}}&newSignature={{alert.series_signature.signature_hash}}&framework={{alertSummary.framework}}" target="_blank">subtests</a>
+            <a ng-show="alert.series_signature.has_subtests" href="{{alert.getSubtestsURL(alertSummary)}}" target="_blank">subtests</a>
```
Comment (wlach) [inline ui/partials/perf/alertsctrl.html]: Use ng-if instead of ng-show here.

**Comment by wlach** [comment]
Sorry, missed a few small nits on the last pass. :)

File: ui/js/controllers/perf/graphs.js (hunk 5 above)
Comment (crosscent) [inline ui/js/controllers/perf/graphs.js]: I did console.log(seriesList), and it was a list of objects with ``hasSubtests`` property.

I think line 50 on the same file parses ``has_subtests`` to the property ``hasSubtests``, or did I not understand what you mean?

Comment (wlach) [inline ui/js/controllers/perf/graphs.js]: This is a change to the test chooser controller. We don't want or need to filter the results based on whether they have a subtests property (in fact there are many cases that don't work properly if we do this, such as when we want to include subtests, or a non-subtest doesn't have subtests itself).

If you just take this change out I think we'll be good to land, the rest of the patch looks great.

**Pull request merged**